addons/payment_paypal/controllers/main.py

```diff
@@ -62,25 +59,24 @@ def paypal_validate_data(self, **post):
         if reference:
             tx = request.env['payment.transaction'].search([('reference', '=', reference)])
         paypal_urls = request.env['payment.acquirer']._get_paypal_urls(tx and tx.acquirer_id.environment or 'prod')
-        pdt_request = bool(new_post.get('amt'))  # check for spefific pdt param
-        if pdt_request:
+        if bool(new_post.get('amt')):  # check for spefific pdt param
             # this means we are in PDT instead of DPN like before
             # fetch the PDT token
-            new_post['at'] = request.env['ir.config_parameter'].sudo().get_param('payment_paypal.pdt_token')
+            new_post['at'] = tx and tx.acquirer_id.pdt_token or ''
             new_post['cmd'] = '_notify-synch'  # command is different in PDT than IPN/DPN
-        validate_url = paypal_urls['paypal_form_url']
-        urequest = urllib2.Request(validate_url, werkzeug.url_encode(new_post))
+        urequest = urllib2.Request(paypal_urls['paypal_form_url'], werkzeug.url_encode(new_post))
         uopen = urllib2.urlopen(urequest)
-        resp = uopen.read()
-        if pdt_request:
-            resp, post = self._parse_pdt_response(resp)
-        if resp == 'VERIFIED' or pdt_request and resp == 'SUCCESS':
+        lines = filter(None, uopen.read().split('\n'))
+        resp = lines.pop(0)
+        if resp == 'SUCCESS': # check resp for wrong PDT token
```

Comment:
I prefer to keep somehow the old method implementation
1/ check if we are in pdf or ipn/dpn
2/ compute response and update post values accordingly
3/ do stuff depending on response like you did

Thanks !